### PR TITLE
Fixes #2847 Getting doc history can go into infinite loop: Backport for 1.3.1.2

### LINKED ISSUE #2847: Getting doc history can go into infinite loop
### Sync Gateway version
1.3.1.2 + possibly others

### Config file

Set revs_limit to `250`

### Steps to reproduce

https://github.com/couchbase/sync_gateway/blob/6979914293befc6ff345d6d7290462d931f26bb3/db/revtree_test.go#L653-L676

### Which versions affected

The original fix (that includes the bug) is the fix for https://github.com/couchbase/sync_gateway/issues/2651.

* 1.3.1 backport: https://github.com/couchbase/sync_gateway/issues/2690
        * 1.4.1.1 backport: https://github.com/couchbase/sync_gateway/commit/a3b4cbd16c740639b6a406f13eb715cc2731c3d1

### Branches to apply changes to

- [x] 1.3.x - https://github.com/couchbase/sync_gateway/pull/2856
- [x] 1.4.x
- [ ] master

### Detect corrupt docs

If walking a rev branch is 2X the revs_limit, we return error to caller

### Auto-fix corrupt docs

- Config flag: fixRevsTreeLoops (on by default, have option to disable)

### Defensive future-proof for additions to 

- If rev generation goes wrong generation when adding a parent, throw error.  Prevent cause in the first place.

### PR DESCRIPTION
**Pull request by tleyden**
Fixes #2847

Here's the PR for the 1.3.1.2 branch.  I created a release 1.3.1.3 base branch based on the 1.3.1.2 release tag, and this PR will pull the changes in from the feature branch.  (should be squashed into single commit)

Once this is reviewed and merged, I'll do the same for 1.4x branch and the master branch

## Changes

- Add unit test with doc that has problematic revs history, that reproduces an infinite loop
- Update `getHistory()` to do cycle detection and return an error (previously named `getValidatedHistory()`
- Update all usages of `getHistory()` to check and propagate the error
- Fix the issue (prevent rev trees with cycles from being created) in two ways:
   - Clean the dangling parents after branch deletes -- tested independently against the repro unit test `TestRevisionPruningLoop `
    - Better handle dangling parents during marshal  -- tested independently against the repro unit test `TestRevisionPruningLoop `

## Tests

- Drone CI
- [build-sync-gateway/247 -- SG + SGA unit tests](http://uberjenkins.sc.couchbase.com/view/Build/job/build-sync-gateway/247/)

### CODE REVIEW

**Comment by tleyden** [comment]
Seeing test failures, looking into this

File: manifest/default.xml [hunk 1]
```diff
@@ -14,7 +14,7 @@
   <remote fetch="git://github.com/kardianos/" name="kardianos"/>
   <remote fetch="git://github.com/coreos/" name="coreos"/>
   <remote fetch="git://github.com/jonboulle/" name="jonboulle"/>
-  <remote fetch="ssh://github.com/couchbaselabs/" name="couchbaselabs_private"/>
+  <remote fetch="ssh://git@github.com/couchbaselabs/" name="couchbaselabs_private"/>
```
Comment (adamcfraser) [inline manifest/default.xml]: Does this need to change?

File: db/revtree.go [hunk 2]
```diff
@@ -128,6 +140,44 @@ func (tree RevTree) UnmarshalJSON(inputjson []byte) (err error) {
 	return
 }
 
+// Validate the revtree.  If it's valid, it will return nil.
+// If invalid, it will return an error with details on why it's invalid.
+func (tree RevTree) Validate() (err error) {
```
Comment (adamcfraser) [inline db/revtree.go]: Can we indicate that this is intended to be test-only in some way?  (maybe ValidateForTest or something) Just so we don't start calling it accidentally in future.

File: db/revtree.go [hunk 3]
```diff
@@ -128,6 +140,44 @@ func (tree RevTree) UnmarshalJSON(inputjson []byte) (err error) {
...
+	if err != nil {
+		return err
+	}
+
+	isValid := (len(rep.Revs) == len(rep.Parents))
+	if !isValid {
+		return fmt.Errorf("rep.Revs length (%d) != rep.Parents length (%d)", len(rep.Revs), len(rep.Parents))
+	}
+
+	log.Printf("rep.Revs length (%d) == rep.Parents length (%d)", len(rep.Revs), len(rep.Parents))
```
Comment (adamcfraser) [inline db/revtree.go]: Even though this is test-only, would be nice to avoid importing log in non-test files.

File: db/revtree.go [hunk 4]
```diff
@@ -534,20 +571,31 @@ func (tree RevTree) RenderGraphvizDot() string {
 	// This function will be called back for every leaf node in tree
 	leafProcessor := func(leaf *RevInfo) {
 
+		// log.Printf("leafProcessor called with leaf: %v", leaf)
+
```
Comment (adamcfraser) [inline db/revtree.go]: Missed comment for cleanup.

**Review by adamcfraser — COMMENTED** [review]
Functionally looks good - just some minor PR cleanup comments.

**Comment by tleyden** [comment]
Addressed PR feedback

**Pull request merged**